File: problems/Java/60. Permutation Sequence.java
class Solution {
    public String getPermutation(int n, int k) {
        int[] fact = new int[n + 1];
        fact[0] = 1;
        List<Integer> nums = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            fact[i] = fact[i - 1] * i;
            nums.add(i);
        }

        StringBuilder res = new StringBuilder();
        k--;

        for (int i = n; i > 0; i--) {
            int idx = k / fact[i - 1];
            k %= fact[i - 1];
            res.append(nums.remove(idx));
        }

        return res.toString();
    }
}

File: problems/C++/60. Permutation Sequence.cpp
class Solution {
public:
    string getPermutation(int n, int k) {
        vector<int> fact(n + 1, 1);
        for (int i = 2; i <= n; ++i) {
            fact[i] = fact[i - 1] * i;
        }

        vector<int> nums(n);
        iota(nums.begin(), nums.end(), 1);

        string res = "";
        k--;

        for (int i = n; i > 0; --i) {
            int idx = k / fact[i - 1];
            k %= fact[i - 1];
            res += to_string(nums[idx]);
            nums.erase(nums.begin() + idx);
        }

        return res;
    }
};